Add new shared/args.js

johnjacobkenny opened the pull request:
Kindly review this PR. It contains only the argument parsing logic for now.

havardh commented on packages/shared/args.js:
```diff
@@ -0,0 +1,30 @@
+        if (arg.indexOf("-") != -1) {
+            const name = getArgName(arg)
+            parsedArgs.named[name] = rest[index + 1];
+            rest.shift();
```
In general, mutating a the object you are iterating over is considered dangerous. And I guess you did a shallow copy of args in order to not mutate the input parameter. So i might think that a standard for loop over args would be better.

johnjacobkenny commented:
Yes, I had wondered whether to go with a simple for loop. Hehe. Please point out anything else, I shall update and send a new PR tomorrow.
________________________________
From: Håvard Wormdal Høiby <notifications@github.com>
Sent: Friday, August 24, 2018 10:33:40 PM
To: havardh/workflow
Cc: Kenny John Jacob; Author
Subject: Re: [havardh/workflow] Add new shared/args.js (#88)


@havardh commented on this pull request.

________________________________

In packages/shared/args.js<https://github.com/havardh/workflow/pull/88#discussion_r212692546>:

> @@ -0,0 +1,30 @@
+function args(args) {
+    const parsedArgs = {
+        named: { },
+        positional: []
+    };
+    const [...rest] = args;
+
+    rest.forEach((arg, index) => {
+        if (arg.indexOf("-") != -1) {
+            const name = getArgName(arg)
+            parsedArgs.named[name] = rest[index + 1];
+            rest.shift();


In general, mutating a the object you are iterating over is considered dangerous. And I guess you did a shallow copy of args in order to not mutate the input parameter. So i might think that a standard for loop over args would be better.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/havardh/workflow/pull/88#pullrequestreview-149383203>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ASWWdJieyeAY-qEHGyY2-UWIerbekdx8ks5uUDHsgaJpZM4WLq-R>.

havardh commented on packages/shared/args.js:
```diff
@@ -0,0 +1,30 @@
+function args(args) {
```
`args(["hello", -c"])` returns `{c: undefined}` I propose this condition should throw an Error.

havardh commented on packages/shared/args.js:
```diff
@@ -0,0 +1,30 @@
+    return arg.slice(1, arg.length);
+}
+
+const input = ['hello', 'world', '-c', 'config', '--filename', 'filename', 'otherargs', 'remaining'];
```
Is your plan to handle the removal of `node cli.js` or `workflow` from `process.argv` prior to calling `args` ? If not the function needs to handle these cases.

johnjacobkenny replied:
I think it makes sense to include the removal of the script name in this file. This could then be used in many other files. Is this simply a matter of checking for the string "node" or "workflow" and splitting as necessary?

havardh replied:
It should be either `node <>.js ...` or `workflow ...`, so at most two arguments. In practice it might actually just be `node <>.js ..` due to the spawn that happens [here](https://github.com/havardh/workflow/blob/master/packages/workflow-cmd/src/cli.js#L15). But I do not think we should commit to that assumption.

johnjacobkenny replied:
I've done an implementation for this, please check if I took the right direction.

havardh commented:
It would be really nice if you add a unit test suite for the parsing functionality :+1: It should go under `shared/test/unit/args_tests.js`

havardh commented on packages/shared/args.js:
```diff
@@ -0,0 +1,30 @@
+}
+
+function getArgName(arg) {
+    if (arg[1] === '-') return arg.slice(2, arg.length);
```
You should be able to run `yarn format` from the root to auto format the files according to project conventions, just making sure you are aware ;)

johnjacobkenny commented:
I've made these changes,

* changed the for loop
* made it throw error
* added tests

Please point out wherever the code looks bad, I'm new to a lot of this.

johnjacobkenny commented:
Should I have rebased from master? Your changes show as deleted

havardh commented:
A simple rebase won't work as the last commit with changes is a "merge commit". So you will need to do some small extra work to make the history correct. If you rebase onto master the last commit will disappear with your changes. So I propose you manually copy the changes you made, do the rebase and then re-add the changes and make a new commit.

havardh commented on packages/shared/test/unit/args_tests.js:
```diff
@@ -0,0 +1,28 @@
+
+    it('it should extract the argument name for word arg', () => {
+        expect(getArgName('--config')).toBe('config');
+    });
```
Add a test case for `getArgName('--some-long-name')` this looks to be failing in the latest commit.

johnjacobkenny replied:
This is implemented.

havardh commented on packages/shared/test/unit/args_tests.js:
```diff
@@ -0,0 +1,28 @@
+import { args, getArgName } from '../../args';
+
+describe('args(args)', () => {
+    it('it should parse the arguments correctly', () => {
```
The second _it_ in `it('it ..` is redundant.

johnjacobkenny replied:
I have given it a try, please see if what I did matches what you had in mind.

havardh commented on packages/shared/test/unit/args_tests.js:
```diff
@@ -0,0 +1,28 @@
+});
+
+describe('getArgName(arg)', () => {
+    it('it should extract the argument name for single character arg', () => {
```
This test states that a '-c' is a single character argument. This is a good condition to have, but it is not enforced in the code. We should make the code throw an Error when the argument is prefixed with a single '-' but the name is longer that one character.

johnjacobkenny replied:
I've added those checks.

havardh commented on packages/shared/args.js:
```diff
@@ -0,0 +1,27 @@
+        if (arg.indexOf("-") != -1) {
+            const name = getArgName(arg);
+            if (args[i + 1] === undefined)
+                throw Error('named argument should have a value');
```
It would be nice to add the name of the argument with missing value to the error message.

johnjacobkenny replied:
I've updated the message.

johnjacobkenny commented:
Does it make sense to include `yarn format` as a git pre-commit hook?

havardh commented:
You are free to do that on in your local repository, these are however usually not commited without tricks like [`shared-git-hooks`](https://github.com/kilianc/shared-git-hooks) or setting the `hooksPath`. I'm happy with this being enforced on the CI server for PRs and for each developer to set up their local environment as they wish. You could go further too say that you would like to have `eslint` and `flow` in that hook too, at that point a commit starts to take a lot of time. You can also set up `prettier` to run on save in your text editor.

As for the PR in general, could you checkout the [commit conventions](https://github.com/havardh/workflow/blob/master/contributions.md#conventions) and complete integrate the functionality into the code base before I continue the review?

johnjacobkenny commented:
Okay

havardh commented:
Hi @johnjacobkenny. I have started working on #89 which rewrites the integration between your argument parser and `workflow`. I have `cherry-picked` your feature into my branch and thus propose closing this PR. Does this sound OK to you? I'm still grateful for your work here, even though I propose not merging your PR. 

If you would like some more _good first issues_ you could take a look at #90 :+1:

johnjacobkenny commented:
Hi @havardh, I'm sorry I got busy with other things. Please continue.. I shall look at #90 when I get more free time. Thank you! (:

havardh commented:
As mentioned above, this was cherry-picked into #133. Thank you for your contribution @johnjacobkenny  :+1:

havardh commented:
You should be able to test this out on [@1.1.1](https://www.npmjs.com/package/workflow)
